<a href="https://colab.research.google.com/github/furrutiav/NLP-competition-2/blob/main/test_gru_competencia2_ner_felipe_urrutia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Instalamos torchtext que nos facilitará la vida en el pre-procesamiento del formato ConLL.
!pip install -U torchtext==0.10.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 7.6 MB 13.9 MB/s 
     |████████████████████████████████| 831.4 MB 2.8 kB/s 
  Attempting uninstall: torch
    Found existing installation: torch 1.11.0+cu113
    Uninstalling torch-1.11.0+cu113:
      Successfully uninstalled torch-1.11.0+cu113
  Attempting uninstall: torchtext
    Found existing installation: torchtext 0.12.0
    Uninstalling torchtext-0.12.0:
      Successfully uninstalled torchtext-0.12.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchvision 0.12.0+cu113 requires torch==1.11.0, but you have torch 1.9.0 which is incompatible.
torchaudio 0.11.0+cu113 requires torch==1.11.0, but you have torch 1.9.0 which is incompatible.


In [ ]:
import torch
from torchtext import data, datasets, legacy


# Garantizar reproducibilidad de los experimentos
SEED = 1234
torch.manual_seed(SEED)
torch.backends.cudnn.deterministic = True

Descargamos los datos de entrenamiento, validación y prueba en nuestro directorio de trabajo

In [ ]:
#%%capture

!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/train.txt -nc # Dataset de Entrenamiento
!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/dev.txt -nc    # Dataset de Validación (Para probar y ajustar el modelo)
!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/test.txt -nc  # Dataset de la Competencia. Estos datos solo contienen los tokens. ¡¡SON LOS QUE DEBEN SER PREDICHOS!!

--2022-06-20 18:58:26--  https://github.com/dccuchile/CC6205/releases/download/v1.0/train.txt
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/196273020/77198f00-c145-11eb-83d1-11e647241ab6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20220620%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220620T185826Z&X-Amz-Expires=300&X-Amz-Signature=d9d5aea2d6707b44b3197c6fce5f08a566da2b3bd9a5573c8490a802d5863a4d&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=196273020&response-content-disposition=attachment%3B%20filename%3Dtrain.txt&response-content-type=application%2Foctet-stream [following]
--2022-06-20 18:58:26--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/196273020/77198f00-c145-11eb-83d1-11e647241ab6?X-Amz-Algorithm=AWS4-H

In [ ]:
import torch
from torchtext import data, datasets, legacy
import pandas as pd
import numpy as np
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
import string

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
# Primer Field: TEXT. Representan los tokens de la secuencia
TEXT = legacy.data.Field(lower=False) 

# Segundo Field: NER_TAGS. Representan los Tags asociados a cada palabra.
NER_TAGS = legacy.data.Field(unk_token=None)
fields = (("text", TEXT), ("nertags", NER_TAGS))


La idea es que usando los fields que definimos antes, le indiquemos a la clase cómo cargar los datasets de prueba, validación y test.

In [ ]:
train_data, valid_data, test_data = legacy.datasets.SequenceTaggingDataset.splits(
    path="./",
    train="train.txt",
    validation="dev.txt",
    test="test.txt",
    fields=fields,
    encoding="utf-8",
    separator=" "
)

In [ ]:
TEXT.build_vocab(train_data)
NER_TAGS.build_vocab(train_data)

In [ ]:
print(f"Tokens únicos en TEXT: {len(TEXT.vocab)}")
print(f"Tokens únicos en NER_TAGS: {len(NER_TAGS.vocab)}")

Tokens únicos en TEXT: 17591
Tokens únicos en NER_TAGS: 12


In [ ]:
#Veamos las posibles etiquetas que hemos cargado:
print(NER_TAGS.vocab.itos)

['<pad>', 'O', 'I-Disease', 'B-Disease', 'I-Body_Part', 'B-Body_Part', 'B-Procedure', 'I-Procedure', 'B-Medication', 'B-Family_Member', 'I-Medication', 'I-Family_Member']


Observen que ademas de los tags NER, tenemos \<pad\>, el cual es generado por el dataloader para cumplir con el padding de cada oración.

Veamos ahora los tokens mas frecuentes y especiales:

In [ ]:
# Seteamos algunas variables que nos serán de utilidad mas adelante...
UNK_IDX = TEXT.vocab.stoi[TEXT.unk_token]
PAD_IDX = TEXT.vocab.stoi[TEXT.pad_token]

PAD_TAG_IDX = NER_TAGS.vocab.stoi[NER_TAGS.pad_token]
O_TAG_IDX = NER_TAGS.vocab.stoi['O']

In [ ]:
BATCH_SIZE = 16  # disminuir si hay problemas de ram.

# Usar cuda si es que está disponible.
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print('Using', device)

# Dividir datos entre entrenamiento y test. Si van a hacer algún sort no puede ser sobre
# el conjunto de testing ya que al hacer sus predicciones sobre el conjunto de test sin etiquetas
# debe conservar el orden original para ser comparado con los golden_labels. 

train_iterator, valid_iterator, test_iterator = legacy.data.BucketIterator.splits(
    (train_data, valid_data, test_data),
    batch_size=BATCH_SIZE,
    device=device,
    sort=False,
)

Using cuda


In [ ]:
!pip install seqeval

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 43 kB 2.5 MB/s 
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16180 sha256=8817cf94c9b29704419060f389795d7ae6a09759a7082d9d1fec66bc87c50d75
  Stored in directory: /root/.cache/pip/wheels/05/96/ee/7cac4e74f3b19e3158dce26a20a1c86b3533c43ec72a549fd7
Successfully built seqeval


In [ ]:
# Definimos las métricas

from seqeval.metrics import f1_score, precision_score, recall_score

def calculate_metrics(preds, y_true, pad_idx=PAD_TAG_IDX, o_idx=O_TAG_IDX, ner_tags=NER_TAGS):
    """
    Calcula precision, recall y f1 de cada batch.
    """

    # Obtener el indice de la clase con probabilidad mayor. (clases)
    y_pred = preds.argmax(dim=1, keepdim=True)

    # filtramos <pad> para calcular los scores.
    mask = [(y_true != pad_idx)]
    y_pred = y_pred[mask]
    y_true = y_true[mask]

    # traemos a la cpu
    y_pred = y_pred.view(-1).to('cpu').numpy()
    y_true = y_true.to('cpu').numpy()
    y_pred = [[ner_tags.vocab.itos[v] for v in y_pred]]
    y_true = [[ner_tags.vocab.itos[v] for v in y_true]]
    
    # calcular scores
    f1 = f1_score(y_true, y_pred, mode='strict')
    precision = precision_score(y_true, y_pred, mode='strict')
    recall = recall_score(y_true, y_pred, mode='strict')

    return precision, recall, f1

In [ ]:
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim


# Definir la red
class NER_RNN(nn.Module):
    def __init__(self, 
                 input_dim, 
                 embedding_dim, 
                 hidden_dim, 
                 output_dim,
                 n_layers, 
                 bidirectional, 
                 dropout, 
                 pad_idx,
                 gru_or_lstm="lstm"):

        super().__init__()
        self.gru_or_lstm = gru_or_lstm

        # Capa de embedding
        self.embedding = nn.Embedding(input_dim,
                                      embedding_dim,
                                      padding_idx=pad_idx)

        # Capa LSTM
        if gru_or_lstm == "lstm":
          self.lstm = nn.LSTM(embedding_dim,
                            hidden_dim,
                            num_layers=n_layers,
                            bidirectional=bidirectional, 
                            dropout = dropout if n_layers > 1 else 0)
        elif gru_or_lstm=="gru":
          self.gru = nn.GRU(embedding_dim, 
                            hidden_dim, 
                            num_layers=n_layers, 
                            bidirectional=bidirectional, 
                            dropout = dropout if n_layers > 1 else 0)

        # Capa de salida
        self.fc = nn.Linear(hidden_dim * 2 if bidirectional else hidden_dim,
                            output_dim)

        # Dropout
        self.dropout = nn.Dropout(dropout)

    def forward(self, text):

        #text = [sent len, batch size]

        # Convertir lo enviado a embedding
        embedded = self.dropout(self.embedding(text))
        
        if self.gru_or_lstm == "lstm":
          outputs, (hidden, cell) = self.lstm(embedded)
        elif self.gru_or_lstm == "gru":
          outputs, hidden = self.gru(embedded)

        #embedded = [sent len, batch size, emb dim]

        # Pasar los embeddings por la rnn (LSTM)

        #output = [sent len, batch size, hid dim * n directions]
        #hidden/cell = [n layers * n directions, batch size, hid dim]

        # Predecir usando la capa de salida.
        predictions = self.fc(self.dropout(outputs))
        #predictions = [sent len, batch size, output dim]

        return predictions

#### **Hiperparámetros de la red**



In [ ]:
# tamaño del vocabulario. recuerden que la entrada son vectores bag of word(one-hot).
INPUT_DIM = len(TEXT.vocab)
EMBEDDING_DIM = 512  # dimensión de los embeddings.
HIDDEN_DIM = 128  # dimensión de la capas LSTM
OUTPUT_DIM = len(NER_TAGS.vocab)  # número de clases

N_LAYERS = 3  # número de capas.
DROPOUT = 0.4
BIDIRECTIONAL = True

# Creamos nuestro modelo.
baseline_model = NER_RNN(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM,
                         N_LAYERS, BIDIRECTIONAL, DROPOUT, PAD_IDX, gru_or_lstm="gru")

baseline_model_name = 'gru'  

In [ ]:
baseline_n_epochs = 10

#### Definimos la función de loss

In [ ]:
# Loss: Cross Entropy
TAG_PAD_IDX = NER_TAGS.vocab.stoi[NER_TAGS.pad_token]
baseline_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX)

--------------------
### Modelo 1

En estas secciones pueden implementar nuevas redes al modificar los hiperparámetros, la cantidad de épocas de entrenamiento, el tamaño de los batches, loss, optimizador, etc... como también definir nuevas arquitecturas de red (mediante la creación de clases nuevas)


Al final de estas, hay 4 variables, las cuales deben setear con los modelos, épocas de entrenamiento, loss y optimizador que deseen probar.


In [ ]:
model_1 = baseline_model
model_name_1 = baseline_model_name
n_epochs_1 = baseline_n_epochs
criterion_1 = baseline_criterion

---------------

### Modelo 2

In [ ]:
# model_2 = masked_model
# model_name_2 = masked_model_name
# n_epochs_2 = baseline_n_epochs
# criterion_2 = masked_criterion

---------------


### Modelo 3

In [ ]:
# model_3 = wo_model
# model_name_3 = wo_model_name
# n_epochs_3 = baseline_n_epochs
# criterion_3= wo_criterion

In [ ]:
model = model_1
model_name = model_name_1
n_epochs = n_epochs_1
criterion= criterion_1

In [ ]:
def init_weights(m):
    # Inicializamos los pesos como aleatorios
    for name, param in m.named_parameters():
        nn.init.normal_(param.data, mean=0, std=0.1) 
        
    # Seteamos como 0 los embeddings de UNK y PAD.
    model.embedding.weight.data[UNK_IDX] = torch.zeros(EMBEDDING_DIM)
    model.embedding.weight.data[PAD_IDX] = torch.zeros(EMBEDDING_DIM)

In [ ]:
model_1.apply(init_weights)

NER_RNN(
  (embedding): Embedding(17591, 512, padding_idx=1)
  (gru): GRU(512, 128, num_layers=3, dropout=0.4, bidirectional=True)
  (fc): Linear(in_features=256, out_features=12, bias=True)
  (dropout): Dropout(p=0.4, inplace=False)
)

In [ ]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f'El modelo actual tiene {count_parameters(model):,} parámetros entrenables.')

El modelo actual tiene 10,095,628 parámetros entrenables.


In [ ]:
# Optimizador
optimizer = optim.Adam(model.parameters())

In [ ]:
# Enviamos el modelo y la loss a cuda (en el caso en que esté disponible)
model = model.to(device)
criterion = criterion.to(device)

In [ ]:
def train(model, iterator, optimizer, criterion):

    epoch_loss = 0
    epoch_precision = 0
    epoch_recall = 0
    epoch_f1 = 0

    model.train()

    # Por cada batch del iterador de la época:
    for batch in iterator:

        # Extraemos el texto y los tags del batch que estamos procesado
        text = batch.text
        tags = batch.nertags

        # Reiniciamos los gradientes calculados en la iteración anterior
        optimizer.zero_grad()

        #text = [sent len, batch size]

        # Predecimos los tags del texto del batch.
        predictions = model(text)

        #predictions = [sent len, batch size, output dim]
        #tags = [sent len, batch size]

        # Reordenamos los datos para calcular la loss
        predictions = predictions.view(-1, predictions.shape[-1])
        tags = tags.view(-1)

        #predictions = [sent len * batch size, output dim]



        # Calculamos el Cross Entropy de las predicciones con respecto a las etiquetas reales
        loss = criterion(predictions, tags)
        
        # Calculamos el accuracy
        precision, recall, f1 = calculate_metrics(predictions, tags, PAD_TAG_IDX, O_TAG_IDX, NER_TAGS)

        # Calculamos los gradientes
        loss.backward()

        # Actualizamos los parámetros de la red
        optimizer.step()

        # Actualizamos el loss y las métricas
        epoch_loss += loss.item()
        epoch_precision += precision
        epoch_recall += recall
        epoch_f1 += f1

    return epoch_loss / len(iterator), epoch_precision / len(
        iterator), epoch_recall / len(iterator), epoch_f1 / len(iterator)

In [ ]:
def evaluate(model, iterator, criterion):

    epoch_loss = 0
    epoch_precision = 0
    epoch_recall = 0
    epoch_f1 = 0

    model.eval()

    # Indicamos que ahora no guardaremos los gradientes
    with torch.no_grad():
        # Por cada batch
        for batch in iterator:

            text = batch.text
            tags = batch.nertags

            # Predecimos
            predictions = model(text)

            predictions = predictions.view(-1, predictions.shape[-1])
            tags = tags.view(-1)

            # Calculamos el Cross Entropy de las predicciones con respecto a las etiquetas reales
            loss = criterion(predictions, tags)

            # Calculamos las métricas
            precision, recall, f1 = calculate_metrics(predictions, tags, PAD_TAG_IDX, O_TAG_IDX, NER_TAGS)

            # Actualizamos el loss y las métricas
            epoch_loss += loss.item()
            epoch_precision += precision
            epoch_recall += recall
            epoch_f1 += f1

    return epoch_loss / len(iterator), epoch_precision / len(
        iterator), epoch_recall / len(iterator), epoch_f1 / len(iterator)

In [ ]:
import time

def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

In [ ]:
best_valid_loss = float('inf')

for epoch in range(n_epochs):

    start_time = time.time()

    # Recuerdo: train_iterator y valid_iterator contienen el dataset dividido en batches.

    # Entrenar
    train_loss, train_precision, train_recall, train_f1 = train(
        model, train_iterator, optimizer, criterion)

    # Evaluar (valid = validación)
    valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
        model, valid_iterator, criterion)

    end_time = time.time()

    epoch_mins, epoch_secs = epoch_time(start_time, end_time)

    # Si obtuvimos mejores resultados, guardamos este modelo en el almacenamiento (para poder cargarlo luego)
    # Si detienen el entrenamiento prematuramente, pueden cargar el modelo en el siguiente recuadro de código.
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), '{}.pt'.format(model_name))
    # Si ya no mejoramos el loss de validación, terminamos de entrenar.

    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(
        f'\tTrain Loss: {train_loss:.3f} | Train f1: {train_f1:.2f} | Train precision: {train_precision:.2f} | Train recall: {train_recall:.2f}'
    )
    print(
        f'\t Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} |  Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
    )

/usr/local/lib/python3.7/dist-packages/seqeval/metrics/sequence_labeling.py:171: UserWarning: <pad> seems not to be NE tag.
  warnings.warn('{} seems not to be NE tag.'.format(chunk))
/usr/local/lib/python3.7/dist-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Epoch: 01 | Epoch Time: 0m 12s
	Train Loss: 0.725 | Train f1: 0.51 | Train precision: 0.63 | Train recall: 0.44
	 Val. Loss: 0.436 |  Val. f1: 0.67 |  Val. precision: 0.71 | Val. recall: 0.65
Epoch: 02 | Epoch Time: 0m 12s
	Train Loss: 0.347 | Train f1: 0.76 | Train precision: 0.79 | Train recall: 0.73
	 Val. Loss: 0.372 |  Val. f1: 0.74 |  Val. precision: 0.77 | Val. recall: 0.73
Epoch: 03 | Epoch Time: 0m 11s
	Train Loss: 0.219 | Train f1: 0.84 | Train precision: 0.85 | Train recall: 0.84
	 Val. Loss: 0.389 |  Val. f1: 0.76 |  Val. precision: 0.77 | Val. recall: 0.76
Epoch: 04 | Epoch Time: 0m 12s
	Train Loss: 0.156 | Train f1: 0.89 | Train precision: 0.89 | Train recall: 0.88
	 Val. Loss: 0.402 |  Val. f1: 0.76 |  Val. precision: 0.78 | Val. recall: 0.75
Epoch: 05 | Epoch Time: 0m 12s
	Train Loss: 0.118 | Train f1: 0.91 | Train precision: 0.92 | Train recall: 0.91
	 Val. Loss: 0.461 |  Val. f1: 0.75 |  Val. precision: 0.74 | Val. recall: 0.76
Epoch: 06 | Epoch Time: 0m 12s
	Train Lo

In [ ]:
# cargar el mejor modelo entrenado.
model.load_state_dict(torch.load('{}.pt'.format(model_name)))

<All keys matched successfully>

In [ ]:
# Limpiar ram de cuda
torch.cuda.empty_cache()

In [ ]:
valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
    model, valid_iterator, criterion)

print(
    f'Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} | Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
)

Val. Loss: 0.372 |  Val. f1: 0.74 | Val. precision: 0.77 | Val. recall: 0.73


Normal:

Val. Loss: 0.398 |  Val. f1: 0.71 | Val. precision: 0.73 | Val. recall: 0.70

GRU: (EMBEDDING_DIM = 256, HIDDEN_DIM = 128, N_LAYERS = 3, DROPOUT = 0.5, BIDIRECTIONAL = False)

Val. Loss: 0.396 |  Val. f1: 0.73 | Val. precision: 0.74 | Val. recall: 0.72

GRU: (EMBEDDING_DIM = 256, HIDDEN_DIM = 128, N_LAYERS = 3, DROPOUT = 0.5, BIDIRECTIONAL = True)

Val. Loss: 0.373 |  Val. f1: 0.77 | Val. precision: 0.79 | Val. recall: 0.76

GRU: (EMBEDDING_DIM = 256, HIDDEN_DIM = 128, N_LAYERS = 1, DROPOUT = 0.5, BIDIRECTIONAL = True)

Val. Loss: 0.348 |  Val. f1: 0.76 | Val. precision: 0.81 | Val. recall: 0.73

GRU: (EMBEDDING_DIM = 128, HIDDEN_DIM = 128, N_LAYERS = 3, DROPOUT = 0.5, BIDIRECTIONAL = True)

Val. Loss: 0.355 |  Val. f1: 0.77 | Val. precision: 0.79 | Val. recall: 0.75

GRU: (EMBEDDING_DIM = 256, HIDDEN_DIM = 128, N_LAYERS = 3, DROPOUT = 0.8, BIDIRECTIONAL = True)

Val. Loss: 0.445 |  Val. f1: 0.73 | Val. precision: 0.75 | Val. recall: 0.72

GRU: (EMBEDDING_DIM = 256, HIDDEN_DIM = 128, N_LAYERS = 3, DROPOUT = 0.2, BIDIRECTIONAL = True)

Val. Loss: 0.338 |  Val. f1: 0.76 | Val. precision: 0.78 | Val. recall: 0.75

GRU: (EMBEDDING_DIM = 256, HIDDEN_DIM = 128, N_LAYERS = 3, DROPOUT = 0.4, BIDIRECTIONAL = True)

Val. Loss: 0.367 |  Val. f1: 0.77 | Val. precision: 0.80 | Val. recall: 0.74

GRU: (EMBEDDING_DIM = 256, HIDDEN_DIM = 128, N_LAYERS = 4, DROPOUT = 0.4, BIDIRECTIONAL = True)

Val. Loss: 0.369 |  Val. f1: 0.76 | Val. precision: 0.77 | Val. recall: 0.75

GRU: (EMBEDDING_DIM = 256, HIDDEN_DIM = 256, N_LAYERS = 3, DROPOUT = 0.4, BIDIRECTIONAL = True)

Val. Loss: 0.380 |  Val. f1: 0.74 | Val. precision: 0.76 | Val. recall: 0.73

GRU: (EMBEDDING_DIM = 512, HIDDEN_DIM = 128, N_LAYERS = 3, DROPOUT = 0.4, BIDIRECTIONAL = True)

Val. Loss: 0.372 |  Val. f1: 0.74 | Val. precision: 0.77 | Val. recall: 0.73